### ISSUE TITLE
Feedback on no separate Avro schema file support

### ISSUE THREAD

**Issue by shoffmeister** [hannes/duckdb_avro#3]
https://duckdb.org/2024/12/09/duckdb-avro-extension.html mentions missing Avro schema file support.

This is somewhat unfortunate, as there _could_ be an exciting use case for ingesting data into DuckDB directly from Apache Kafka: it is not uncommon at all in the Apache Kafka world to encode record values as (more or less) raw binary Avro.

https://docs.confluent.io/platform/current/schema-registry/fundamentals/serdes-develop/serdes-avro.html#serdes-and-formatter-avro describes
* a five byte (1 + 4) custom magic header followed by
* raw binary Avro

The custom magic header can then be used to retrieve data from a schema registry, where the 4 byte word points to a schema registry entry - which has the schema.

While the above implies using a schema registry, simply dropping the magic header and reading the raw binary Avro is possible, too.

What I am trying to get at is that having, for instance, importer parameters such as
* map from schema id to _local_ schema file (and make use of the custom magic header)
* immediate schema provisioning to a local file (and read only the raw binary)
Avro ingest support could become more general and more powerful.

My personal immediate use case for all of that would be a "Kafka extension" offering the ability to "attach to topic", performing background continuous streaming of complete topic content (Kafka records) into a read-only DuckDB table.

Attaching to _multiple_ topics then would allow for statically executed, bounded (i.e. non-streaming) cross-table queries on (potentially) large amounts of data, very much suitable for building ad-hoc analysis on, say, development Kafka systems, or for operational purposes. 

For context, I am considering to write custom code for that ingest into DuckDB in custom Python code, with significant challenges in performance anticipated.

**Comment by jkolash** [comment]
Yes being able to pass in a schema is very desirable

**Comment by jkolash** [comment]
a temporary workaround could be to collect records from kafka in memory and then write an in memory avro "file" that we could pass into duckdb so we are just concatenating bytes in an outputstream in the glue code that we can pass to duckdb.

**Comment by jkolash** [comment]
Is it possible to create a virtual filename via the python and jdbc drivers? Does it need to be seekable or can you provide a simple input stream that supports .read() operations but not seek.

I don't se anything for java in particular
https://duckdb.org/docs/api/java.html

nor do I see one for python
https://duckdb.org/docs/api/python/overview

**Comment by hannes** [comment]
In Python you could try `fsspec`, that's supported in DuckDB

**Comment by hannes** [comment]
Here is a hacky script to generate an Avro file directly from Kafka messages: https://bsky.app/profile/hannes.muehleisen.org/post/3ldo6xtwits25
https://gist.github.com/hannes/7cf7743560333890ad9b3b8e0d9acfc2

**Comment by shoffmeister** [comment]
Aggregating the complete Kafka topic into an .avro file, appending _raw_ Avro, is an interesting strategy! 

On my end, "changelog" topics dominate; I'd have to actually keep more of the Kafka record around, to be able to run "last_value(record.value) over (partition by record.partition, record.key, order by record.offset asc)" analytical queries, to take into account the ordering of the Kafka (log record) stream.

**Comment by jkolash** [comment]
For your information 
See also: https://avro.apache.org/docs/1.12.0/specification/#schema-fingerprints

the confluent schema registry id is based on a monotonic id that gets incremented per schema revision vs a content addressable schema Id based on the schema contents itself 

See the below for another decoder

https://github.com/apache/avro/blob/da39228/lang/java/avro/src/main/java/org/apache/avro/message/BinaryMessageDecoder.java#L146

**Comment by hannes** [comment]
We could add a `avro_decode(binary_blob, schema_string)` function, would that work for you?

**Comment by jkolash** [comment]
Yeah that matches ```from_avro``` in spark https://spark.apache.org/docs/latest/sql-data-sources-avro.html#to_avro-and-from_avro assuming the schema_string needs to be a literal.

It could be up to the application to take care of demultiplexing binary_blobs with different schema Ids. especially when a schema is evolved to a newer version, say by adding a column.

**Comment by hannes** [comment]
Yes the schema would have to be a constant since we need to know at bind time what the output type is going to be. But thanks for the Spark pointer, will add.
